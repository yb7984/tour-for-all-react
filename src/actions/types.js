/** action types for reducers */
export const AUTH_SET_TOKEN = "AUTH_SET_TOKEN";
export const AUTH_SET_USER = "AUTH_SET_USER";
export const AUTH_RESET = "AUTH_RESET";

export const LOAD_USER = "LOAD_USER";

export const LOAD_TOUR = "LOAD_TOUR";

export const RESET_ALL = "RESET_ALL";