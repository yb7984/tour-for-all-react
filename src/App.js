import './App.css';
import Navbar from './Navbar';
import Routes from './Routes';
import { Container } from '@material-ui/core';
import Footer from './Footer';

function App() {
    return (
        <Container>
            <Navbar />
            <Routes />
            <Footer />
        </Container>
    );
}

export default App;
