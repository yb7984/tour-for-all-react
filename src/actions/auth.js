import User from "../models/user";
import { AUTH_SET_TOKEN, RESET_ALL } from "../actions/types";
import { clearClockSockets } from "./clock";

/** actions for auth */

function login(username, password, setLoading, setError) {
    return async function (dispatch) {
        dispatch(resetAll());
        clearClockSockets();        // close all socket connections

        setLoading && setLoading(true);

        try {
            const res = await User.login(username, password);

            if (res.token) {
                dispatch(gotToken(res.token));
            }
        } catch (error) {
            setError && setError(error);
        }
        setLoading && setLoading(false);
    }
}

function logout() {
    return async function (dispatch) {
        dispatch(resetAll());
        clearClockSockets();        // close all socket connections
    }
}


function signup(data, setLoading, setError) {
    return async function (dispatch) {
        dispatch(resetAll());
        clearClockSockets();        // close all socket connections

        setLoading && setLoading(true);
        try {
            const res = await User.register(data);

            if (res.token) {
                dispatch(gotToken(res.token));
            }
        } catch (error) {
            if (setError) {
                setError && setError(error);
            }
        }
        setLoading && setLoading(false);
    }
}

function gotToken(token) {
    return {
        type: AUTH_SET_TOKEN,
        payload: token
    };
}

function resetAll() {
    return { type: RESET_ALL };
}

export {
    login,
    logout,
    signup
};