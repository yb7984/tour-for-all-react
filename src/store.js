import  { composeWithDevTools} from "redux-devtools-extension";
import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import autoMergeLevel2 from "redux-persist/lib/stateReconciler/autoMergeLevel2";

import thunk from "redux-thunk";
import root from "./reducers/root";
import { createStore, applyMiddleware } from "redux";

/** Persist redux store */

const persistConfig = {
  key: "root",          // persist the root reducer
  storage,              // using localStorage
  stateReconciler: autoMergeLevel2  // shallow merges two levels
};

const persistedReducer = persistReducer(persistConfig, root);


export const store = createStore(
  persistedReducer,
  composeWithDevTools(
    applyMiddleware(thunk),     // apply thunk
  )
);

export const persistedStore = persistStore(store);