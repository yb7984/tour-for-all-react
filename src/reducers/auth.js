import { AUTH_SET_TOKEN, AUTH_RESET } from "../actions/types";
import { ROLE_USER } from "../models/role";
import jwt from "jsonwebtoken";

/** reducer for tours */

const INITIAL_STATE = {
    token: null,
    username: null,
    role: ROLE_USER
};


function auth(state = INITIAL_STATE, action) {
    switch (action.type) {
        case AUTH_RESET:
            return { ...INITIAL_STATE };
        case AUTH_SET_TOKEN:
            const token = action.payload;
            const { username, role } = jwt.decode(token);
            return {
                token,
                username,
                role
            };
        default:
            return state;
    }
}

export default auth;