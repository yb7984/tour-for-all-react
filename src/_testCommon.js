"use strict";
import { render as originRender } from '@testing-library/react';
import jwt from "jsonwebtoken";

import '@fontsource/roboto';

import CssBaseline from '@material-ui/core/CssBaseline';
import { ThemeProvider } from '@material-ui/core/styles';

import theme from './theme';
import { Provider } from "react-redux";
import { store } from "./store";
import '@fontsource/roboto';
import User from './models/user';
import Tour from './models/tour';
import { ROLE_ADMIN, ROLE_USER } from './models/role';
import { TOUR_STATUS_PUBLIC } from '../../tour-for-all-express/models/tourStatus';

/**
 * return signed JWT from user data.
 * @param {Object} user 
 * @returns token
 */
function createToken(user) {
    console.assert(user.role !== undefined,
        "createToken passed user without role property");

    let payload = {
        username: user.username,
        role: user.role || ROLE_USER,
    };

    return jwt.sign(payload, "SECRET_KEY");
}


export function render(child) {
    return originRender(
        <Provider store={store}>
            <ThemeProvider theme={theme}>
                <CssBaseline />
                {child}
            </ThemeProvider>
        </Provider>
    );
}

export const u1 = new User({
    username: "u1",
    password: 'password1',
    firstName: "U1F",
    lastName: "U1L",
    email: "u1@email.com",
    role: ROLE_ADMIN,
    isActive: true
});
export const u2 = new User({
    username: "u2",
    password: 'password2',
    firstName: "U2F",
    lastName: "U2L",
    email: "u2@email.com",
    role: ROLE_USER,
    isActive: true
});
export const t1 = new Tour({
    title: "t-1",
    creator: "u1",
    price: 10,
    entryFee: 2,
    start: new Date(Date.now() + 24 * 60 * 60 * 1000),
    status: TOUR_STATUS_PUBLIC,
    isActive: true
});
export const t2 = new Tour({
    title: "t-2",
    creator: "u2",
    price: 100,
    entryFee: 0,
    start: new Date(Date.now() - 4 * 60 * 60 * 1000),
    status: TOUR_STATUS_PUBLIC,
    isActive: true
});


export const u1Token = createToken({ username: "u1", role: ROLE_ADMIN });
export const u2Token = createToken({ username: "u2", role: ROLE_USER });