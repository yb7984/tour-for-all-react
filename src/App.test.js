import App from './App';
import { render, u1Token, t1, u1 } from "./_testCommon";
import axiosMock from 'axios';
import { MemoryRouter } from 'react-router';
import TourForAllAPI from "./api";
import { fireEvent, queryByText } from '@testing-library/dom';
import { store } from './store';
import { resetAll } from './actions/auth';

const originConsoleError = console.error;

beforeEach(() => {
    console.error = jest.fn();
    store.dispatch(resetAll());     //clear everything in redux store
});

afterEach(() => {
    console.error = originConsoleError;
});

describe("App can render", function () {
    it('renders without login', () => {
        Object.defineProperty(TourForAllAPI, 'token', {
            get: jest.fn(() => '')
        });

        const data = {
            tours: [],
            page: 1,
            total: 0,
            perPage: 3
        }
        axiosMock.mockResolvedValue(data);

        render(<MemoryRouter initialEntries={['/']}><App /></MemoryRouter>);
        render(<MemoryRouter initialEntries={['/tours']}><App /></MemoryRouter>);
        render(<MemoryRouter initialEntries={['/login']}><App /></MemoryRouter>);
        render(<MemoryRouter initialEntries={['/signup']}><App /></MemoryRouter>);
    });


    it('test login', async () => {
        Object.defineProperty(TourForAllAPI, 'token', {
            get: jest.fn(() => '')
        });

        axiosMock.mockImplementation(({ url }) => {

            if (url.indexOf("/token") !== -1) {
                return {
                    status: 200,
                    data: {
                        token: u1Token
                    }
                };
            }

            if (url.indexOf("/users") !== -1) {
                return {
                    status: 200,
                    data: {
                        user: { ...u1 }
                    }
                };
            }


            if (url.indexOf("/tours") !== -1) {
                return {
                    status: 200,
                    data: {
                        tours: [{ ...t1 }],
                        total: 1,
                        perPage: 3,
                        page: 1
                    }
                };
            }

        });

        const { asFragment, findByText, queryAllByText, queryByText, getByTestId, getByPlaceholderText } = render(<MemoryRouter initialEntries={['/login']}><App /></MemoryRouter>);

        // expect(asFragment()).toMatchSnapshot();

        expect(getByPlaceholderText("Enter username")).toBeInTheDocument();
        let logins = queryAllByText("Login");
        logins.forEach(login => {
            expect(login).toBeInTheDocument();
        });

        let signups = queryAllByText("Sign up");
        signups.forEach(signup => {
            expect(signup).toBeInTheDocument();
        });
        expect(queryByText("Logout")).not.toBeInTheDocument();
        expect(queryByText("Profile")).not.toBeInTheDocument();
        expect(queryByText("My Tours")).not.toBeInTheDocument();
        expect(queryByText("Joined")).not.toBeInTheDocument();
        expect(queryByText("Favorites")).not.toBeInTheDocument();

        fireEvent.change(getByPlaceholderText("Enter username"), { target: { value: "u1" } });
        fireEvent.change(getByPlaceholderText("Enter password"), { target: { value: "password" } });

        //login
        fireEvent.click(getByTestId("login-button"));

        await findByText("Joined Tournaments");
        expect(queryByText("U1F U1L", { exact: false })).toBeInTheDocument();     //show the name of the use 
        expect(queryByText("Logout")).toBeInTheDocument();
        expect(queryByText("Login")).not.toBeInTheDocument();
        expect(queryByText("Sign up")).not.toBeInTheDocument();
        expect(queryByText("Profile")).toBeInTheDocument();
        expect(queryByText("My Tours")).toBeInTheDocument();
        expect(queryByText("Joined")).toBeInTheDocument();
        expect(queryByText("Favorites")).toBeInTheDocument();
    });



    it('test signup', async () => {
        Object.defineProperty(TourForAllAPI, 'token', {
            get: jest.fn(() => '')
        });

        axiosMock.mockImplementation(({ url }) => {

            if (url.indexOf("/register") !== -1) {
                return {
                    status: 201,
                    data: {
                        token: u1Token
                    }
                };
            }

            if (url.indexOf("/users") !== -1) {
                return {
                    status: 200,
                    data: {
                        user: { ...u1 }
                    }
                };
            }


            if (url.indexOf("/tours") !== -1) {
                return {
                    status: 200,
                    data: {
                        tours: [{ ...t1 }],
                        total: 1,
                        perPage: 3,
                        page: 1
                    }
                };
            }

        });

        const { asFragment, findByText, queryByText, getByTestId, getByPlaceholderText } =
            render(<MemoryRouter initialEntries={['/signup']}><App /></MemoryRouter>);

        // expect(asFragment()).toMatchSnapshot();


        expect(queryByText("User Sign Up", { exact: false })).toBeInTheDocument();

        fireEvent.change(getByPlaceholderText("Enter username"), { target: { value: "u1" } });
        fireEvent.change(getByPlaceholderText("Enter password"), { target: { value: "password" } });
        fireEvent.change(getByPlaceholderText("Enter first name"), { target: { value: u1.firstName } });
        fireEvent.change(getByPlaceholderText("Enter last name"), { target: { value: u1.lastName } });
        fireEvent.change(getByPlaceholderText("Enter email"), { target: { value: u1.email } });

        //login
        fireEvent.click(getByTestId("signup-button"));

        await findByText("Joined Tournaments");
        expect(queryByText("U1F U1L", { exact: false })).toBeInTheDocument();     //show the name of the use 
    });
});