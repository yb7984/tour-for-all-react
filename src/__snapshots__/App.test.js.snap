// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App can render test login 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <div
          class="MuiBox-root MuiBox-root-57"
        >
          <button
            aria-label="menu"
            class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-42 MuiIconButton-colorInherit MuiIconButton-edgeStart"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <h6
          class="MuiTypography-root makeStyles-title-43 MuiTypography-h6"
        >
          <a
            aria-current="page"
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-titleButton-44 MuiButton-colorInherit active"
            href="/"
            role="button"
            tabindex="0"
          >
            <span
              class="MuiButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-titleIcon-45"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <svg>
                  poker1.svg
                </svg>
              </svg>
              Tour for All
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
        </h6>
        <div
          class="MuiBox-root MuiBox-root-58"
        >
          <div
            class="makeStyles-menuHorizontal-53"
          >
            <ul
              class="MuiList-root makeStyles-listHorizontal-54 MuiList-padding"
            >
              <a
                aria-current="page"
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-listItemHorizontal-56 MuiListItem-gutters MuiListItem-button active"
                href="/"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Home
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
              <a
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-listItemHorizontal-56 MuiListItem-gutters MuiListItem-button"
                href="/tours"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Tournaments
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </ul>
            <hr
              class="MuiDivider-root"
            />
            <ul
              class="MuiList-root makeStyles-listHorizontal-54 MuiList-padding"
            >
              <a
                aria-current="page"
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-listItemHorizontal-56 MuiListItem-gutters MuiListItem-button active"
                href="/login"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Login
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
              <a
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-listItemHorizontal-56 MuiListItem-gutters MuiListItem-button"
                href="/signup"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Sign up
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </ul>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root makeStyles-backdrop-59"
        style="opacity: 0; visibility: hidden;"
      >
        <div
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </div>
    </header>
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <form
        class="makeStyles-root-60"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <h3
              class="MuiTypography-root MuiTypography-h3"
            >
              User Login
            </h3>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-colorError"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused Mui-required Mui-required"
                data-shrink="true"
                for="username"
                id="username-label"
              >
                Username
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  id="username"
                  name="username"
                  placeholder="Enter username"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-61 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-63 PrivateNotchedOutline-legendNotched-64"
                  >
                    <span>
                      Username *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="password"
                id="password-label"
              >
                Password
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  id="password"
                  name="password"
                  placeholder="Enter password"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-61 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-63"
                  >
                    <span>
                      Password *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              data-testid="login-button"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-label"
              >
                Login
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6"
            >
              No account? just 
              <a
                href="/signup"
              >
                Sign up
              </a>
               here.
            </h6>
          </div>
        </div>
      </form>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root makeStyles-backdrop-59"
        style="opacity: 0; visibility: hidden;"
      >
        <div
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App can render test signup 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <div
          class="MuiBox-root MuiBox-root-93"
        >
          <button
            aria-label="menu"
            class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-78 MuiIconButton-colorInherit MuiIconButton-edgeStart"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <h6
          class="MuiTypography-root makeStyles-title-79 MuiTypography-h6"
        >
          <a
            aria-current="page"
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-titleButton-80 MuiButton-colorInherit active"
            href="/"
            role="button"
            tabindex="0"
          >
            <span
              class="MuiButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-titleIcon-81"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <svg>
                  poker1.svg
                </svg>
              </svg>
              Tour for All
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
        </h6>
        <div
          class="MuiBox-root MuiBox-root-94"
        >
          <div
            class="makeStyles-menuHorizontal-89"
          >
            <ul
              class="MuiList-root makeStyles-listHorizontal-90 MuiList-padding"
            >
              <a
                aria-current="page"
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-listItemHorizontal-92 MuiListItem-gutters MuiListItem-button active"
                href="/"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Home
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
              <a
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-listItemHorizontal-92 MuiListItem-gutters MuiListItem-button"
                href="/tours"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Tournaments
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </ul>
            <hr
              class="MuiDivider-root"
            />
            <ul
              class="MuiList-root makeStyles-listHorizontal-90 MuiList-padding"
            >
              <a
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-listItemHorizontal-92 MuiListItem-gutters MuiListItem-button"
                href="/login"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Login
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
              <a
                aria-current="page"
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-listItemHorizontal-92 MuiListItem-gutters MuiListItem-button active"
                href="/signup"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Sign up
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </ul>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root makeStyles-backdrop-95"
        style="opacity: 0; visibility: hidden;"
      >
        <div
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </div>
    </header>
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <form
        class="makeStyles-root-96"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <h3
              class="MuiTypography-root MuiTypography-h3"
            >
              User Sign Up
            </h3>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused Mui-required Mui-required"
                data-shrink="true"
                for="username"
                id="username-label"
              >
                Username
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  id="username"
                  maxlength="30"
                  name="username"
                  placeholder="Enter username"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-97 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-99 PrivateNotchedOutline-legendNotched-100"
                  >
                    <span>
                      Username *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="password"
                id="password-label"
              >
                Password
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  id="password"
                  maxlength="30"
                  minlength="5"
                  name="password"
                  placeholder="Enter password"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-97 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-99"
                  >
                    <span>
                      Password *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="firstName"
                id="firstName-label"
              >
                First Name
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  id="firstName"
                  maxlength="30"
                  name="firstName"
                  placeholder="Enter first name"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-97 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-99"
                  >
                    <span>
                      First Name *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="lastName"
                id="lastName-label"
              >
                Last Name
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  id="lastName"
                  maxlength="30"
                  name="lastName"
                  placeholder="Enter last name"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-97 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-99"
                  >
                    <span>
                      Last Name *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="email"
                id="email-label"
              >
                Email
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  id="email"
                  maxlength="50"
                  name="email"
                  placeholder="Enter email"
                  required=""
                  type="email"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-97 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-99"
                  >
                    <span>
                      Email *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              data-testid="signup-button"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-label"
              >
                Sign Up
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </form>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root makeStyles-backdrop-95"
        style="opacity: 0; visibility: hidden;"
      >
        <div
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
